### ISSUE TITLE
Runtime Performance

### ISSUE THREAD

**Issue by PedroCastro** [zju3dv/pvnet#31]
Hi there! Thank you for making this available and congrats on getting accepted into CVPR!

I was reading your paper and I was curious on how you made your inference speed so fast with such big inputs. Using this implementation along with your pre-trained models I re-measured each component speed and found it to be different from what you detailed in the paper. (used the technique described [here](https://discuss.pytorch.org/t/how-to-measure-time-in-pytorch/26964/2))

I then wrapped the [net call](https://github.com/zju3dv/pvnet/blob/a85723fd2c92cbb3f6bb1711a808ae6f3022e17a/tools/train_linemod.py#L193), the [RANSAC call](https://github.com/zju3dv/pvnet/blob/a85723fd2c92cbb3f6bb1711a808ae6f3022e17a/tools/train_linemod.py#L201) and from line 167 to the [PnP call](https://github.com/zju3dv/pvnet/blob/a85723fd2c92cbb3f6bb1711a808ae6f3022e17a/lib/utils/evaluation_utils.py#L188) with the previously mentioned timing code. Using a 1080Ti and a i5-8500 CPU 3.00GHz, when calling "python tools/train_linemod.py --cfg_file configs/linemod_train.json --linemod_cls ape --test_model --use_uncertainty_pnp", I got the following times

- Forward Pass: ~27ms
- RANSAC: ~14ms
- PnP: ~1.6ms

Both RANSAC and the forward pass run on the GPU so I expected the processing time to be the same. What am I doing something wrong? The whole evaluation (with metric measuring and loading) ran at 11fps.

Thanks in advance!

**Comment by pengsida** [comment]
How about testing several times and taking the average time?
Your RANSAC and PnP performance is similar to us, but the forward pass has a large gap.
Our running time does not include the metric measuring.
The reason our method is real-time is that our backbone is Resnet-18.

![image](https://user-images.githubusercontent.com/11582122/59270401-f6290a80-8c83-11e9-947c-f905d37fa56c.png)

**Comment by PedroCastro** [comment]
Thanks for the answer!

Well, I am already taking the average among all test images for the Ape object. Do you mean taking the avg of multiple objects? That shouldn't change the inference speed so I am not sure what I can measure differently.

I agree that Resnet-18 is quite fast. But for example, [this pre-print](https://arxiv.org/pdf/1902.11020.pdf) which also uses Resnet-18 with a much smaller input (320x240) claims their forward propagation takes 11ms, on a Titan X.

Am I measuring a different call than what you guys did? How did you measure it?

Thanks again for the conversation!

**Comment by pengsida** [comment]
Their architecture has three decoders.
I will ask my partner to double check our forward pass time.

![image](https://user-images.githubusercontent.com/11582122/59283135-a276eb00-8c9c-11e9-89e7-4714d039ad1f.png)

**Comment by pengsida** [comment]
Does your forward pass includes the data loading?

**Comment by PedroCastro** [comment]
Oh @pengsida, you are right! But even if you divide the full time by 3 (which would divide the time on the encoder part as well) it would match your speed with half the resolution.

And no, it doesnt take the data loading into account! I synchronize cuda calls before and after the forward pass.
My code is like this:
`
    torch.cuda.synchronize()
    start.record()
    seg_pred, vertex_pred, loss_seg, loss_vertex, precision, recall = net(image, mask, vertex, vertex_weights)
    torch.cuda.synchronize()
    end.record()
    time_accumulator += [start.elapsed_time(end)]
`
and then I avg the time at the end.

I think your method does run at 25fps but the way your time is descritized in the paper might not be right.

**Comment by pengsida** [comment]
Probably.
A double check is needed.
It is indeed hard to directly measure the cuda running time.
Whatever, our approach is real-time in our use.

**Comment by PedroCastro** [comment]
Yes, good job and good paper ;)

Let me know what you find!

**Comment by liuyuan-pal** [comment]
Hi PedroCastro,
The former testing code didn't use synchronize, which caused incorrect time records for each stage.
I add torch.cuda.synchronize() and find the forward pass indeed costs about 27 ms and the RANSAC costs about 14ms. 
Thank you for pointing out the mistakes!
